# CPE 486/586 – Homework 3
## Optimization and Linear Regression

Name: Julio Cerda
UAH Email: jcc0065@uah.edu
Unique Keyword: bluejet
GitHub URL: [Your GitHub URL]
GitHub Commit Hash: [add at the end]

## Question 1 – Automatic Differentiation and Expression Graph

Given:

$$
f(w,x,y,z)=wx+y^2z
$$

where

$$
w=1,\quad x=2,\quad y=3,\quad z=4
$$

Break the function into intermediate operations:

$$
a=wx
$$

$$
b=y^2
$$

$$
c=bz
$$

$$
f=a+c
$$

w ----\
       * ---- a ----\
x ----/              \
                       * ---- f
y ----\              /
       * ---- b ----*
y ----/            /
                  /
z ---------------/

Using the given values:

$$
a=wx=(1)(2)=2
$$

$$
b=y^2=(3)^2=9
$$

$$
c=bz=(9)(4)=36
$$

$$
f=a+c=2+36=38
$$

For the derivative with respect to $w$, follow the path from $w$ to $f$:

$$
\frac{\partial a}{\partial w}=x=2
$$

and

$$
\frac{\partial f}{\partial a}=1
$$

Therefore,

$$
\frac{\partial f}{\partial w}
=
\frac{\partial f}{\partial a}
\frac{\partial a}{\partial w}
=
(1)(2)=2
$$

Thus,

$$
\boxed{\frac{\partial f}{\partial w}=2}
$$

## Question 2 – Unconstrained Optimization

### 2.1

Given:

$$
f(x)=x^2-4x+4
$$

Take the derivative:

$$
f'(x)=2x-4
$$

Set the derivative equal to zero to find the critical point:

$$
2x-4=0
$$

$$
2x=4
$$

$$
x=2
$$

Therefore, the critical point is:

$$
\boxed{x=2}
$$

To determine whether this point is a minimum or maximum, take the second derivative:

$$
f''(x)=2
$$

Since:

$$
f''(x)>0
$$

the critical point at $x=2$ is a local minimum.

Now evaluate the function at $x=2$:

$$
f(2)=(2)^2-4(2)+4
$$

$$
f(2)=4-8+4=0
$$

Therefore:

$$
\boxed{\text{Minimum value}=0}
$$

### 2.2

Given:

$$
f(x)=x^3-3x^2
$$

Take the first derivative:

$$
f'(x)=3x^2-6x
$$

Factor:

$$
f'(x)=3x(x-2)
$$

Set the derivative equal to zero:

$$
3x(x-2)=0
$$

Therefore, the critical points are:

$$
\boxed{x=0}
$$

and

$$
\boxed{x=2}
$$

Now take the second derivative:

$$
f''(x)=6x-6
$$

Evaluate at $x=0$:

$$
f''(0)=6(0)-6=-6
$$

Since:

$$
f''(0)<0
$$

$x=0$ is a local maximum.

Now evaluate the function:

$$
f(0)=(0)^3-3(0)^2=0
$$

Therefore:

$$
\boxed{\text{Local maximum at }x=0,\quad f(0)=0}
$$

Evaluate the second derivative at $x=2$:

$$
f''(2)=6(2)-6=6
$$

Since:

$$
f''(2)>0
$$

$x=2$ is a local minimum.

Now evaluate the function:

$$
f(2)=(2)^3-3(2)^2
$$

$$
f(2)=8-12=-4
$$

Therefore:

$$
\boxed{\text{Local minimum at }x=2,\quad f(2)=-4}
$$

### 2.3

Given:

$$
f(x,y)=x^2-y^2
$$

Compute the gradient:

$$
\nabla f=
\left[
\frac{\partial f}{\partial x},
\frac{\partial f}{\partial y}
\right]
$$

The partial derivatives are:

$$
\frac{\partial f}{\partial x}=2x
$$

and

$$
\frac{\partial f}{\partial y}=-2y
$$

Therefore:

$$
\nabla f=
\left[
2x,-2y
\right]
$$

Set the gradient equal to zero to find the critical point:

$$
2x=0
$$

$$
-2y=0
$$

Therefore:

$$
x=0,\quad y=0
$$

So the critical point is:

$$
\boxed{(0,0)}
$$

To classify the point, consider the function along different directions.

Along the x-axis, where $y=0$:

$$
f(x,0)=x^2
$$

which is positive for $x\neq0$.

Along the y-axis, where $x=0$:

$$
f(0,y)=-y^2
$$

which is negative for $y\neq0$.

Since the function increases in one direction and decreases in another direction around $(0,0)$, the critical point is a saddle point.

Therefore:

$$
\boxed{(0,0)\text{ is a saddle point}}
$$

## Question 3 – Constrained Optimization Using Lagrange Multipliers

Given:

$$
f(x,y,z)=x^2+y^2+z^2
$$

subject to the constraint:

$$
x+2y+z=6
$$

### 3.1 Set up the Lagrangian

Rewrite the constraint as:

$$
x+2y+z-6=0
$$

The Lagrangian is:

$$
\mathcal{L}(x,y,z,\lambda)
=
x^2+y^2+z^2+\lambda(x+2y+z-6)
$$

### 3.2 Derive the system of equations

Take the partial derivatives and set them equal to zero:

$$
\frac{\partial \mathcal{L}}{\partial x}
=
2x+\lambda=0
$$

$$
\frac{\partial \mathcal{L}}{\partial y}
=
2y+2\lambda=0
$$

$$
\frac{\partial \mathcal{L}}{\partial z}
=
2z+\lambda=0
$$

$$
\frac{\partial \mathcal{L}}{\partial \lambda}
=
x+2y+z-6=0
$$

### 3.3 Solve for the critical point

From:

$$
2x+\lambda=0
$$

$$
x=-\frac{\lambda}{2}
$$

From:

$$
2y+2\lambda=0
$$

$$
y=-\lambda
$$

From:

$$
2z+\lambda=0
$$

$$
z=-\frac{\lambda}{2}
$$

Substitute into the constraint:

$$
x+2y+z=6
$$

$$
-\frac{\lambda}{2}
+2(-\lambda)
-\frac{\lambda}{2}
=6
$$

$$
-3\lambda=6
$$

$$
\lambda=-2
$$

Therefore:

$$
x=1,\quad y=2,\quad z=1
$$

So the critical point is:

$$
\boxed{(1,2,1)}
$$

### 3.4 Classify the extremum

The objective function is:

$$
f(x,y,z)=x^2+y^2+z^2
$$

This function is convex because it is the sum of squared terms. Therefore, the critical point found using the Lagrange multiplier method is a minimum subject to the constraint.

At the critical point:

$$
(x,y,z)=(1,2,1)
$$

the function value is:

$$
f(1,2,1)=1^2+2^2+1^2
$$

$$
f(1,2,1)=1+4+1=6
$$

Therefore:

$$
\boxed{\text{Constrained minimum at }(1,2,1)}
$$

with minimum value:

$$
\boxed{f_{\min}=6}
$$

### 3.5 Explain the role of the constraint

The constraint

$$
x+2y+z=6
$$

restricts the possible values of $x$, $y$, and $z$ to only the points that satisfy this equation.

Without the constraint, the function

$$
f(x,y,z)=x^2+y^2+z^2
$$

would have its minimum at:

$$
(0,0,0)
$$

However, $(0,0,0)$ does not satisfy the constraint.

Therefore, the Lagrange multiplier method finds the point on the constraint surface that gives the smallest possible value of the objective function.

For this problem, that point is:

$$
\boxed{(1,2,1)}
$$

with:

$$
\boxed{f_{\min}=6}
$$

## Question 4 – Optimizing a Function Using Stochastic Gradient Descent

Given:

$$
J(w_1)=(w_1-5)^4+(w_1-3)^2+4
$$

with:

$$
\eta=0.1
$$

and:

$$
w_1^{(0)}=2.0
$$

The derivative is:

$$
\frac{dJ}{dw_1}
=
4(w_1-5)^3+2(w_1-3)
$$

The gradient descent update equation is:

$$
w_1^{(k+1)}
=
w_1^{(k)}
-
0.1\left[4(w_1^{(k)}-5)^3+2(w_1^{(k)}-3)\right]
$$

### Iteration 1

$$
\frac{dJ}{dw_1}
=
4(2-5)^3+2(2-3)
=-110
$$

$$
w_1^{(1)}
=
2-0.1(-110)
=13
$$

### Iteration 2

$$
\frac{dJ}{dw_1}
=
4(13-5)^3+2(13-3)
=2068
$$

$$
w_1^{(2)}
=
13-0.1(2068)
=-193.8
$$

### Iteration 3

$$
w_1^{(3)}
\approx
3.1425904688\times10^6
$$

### Iteration 4

$$
w_1^{(4)}
\approx
-1.2414272820484\times10^{19}
$$

### Iteration 5

$$
w_1^{(5)}
\approx
7.652861390717\times10^{56}
$$

The magnitude of the updates grows rapidly, so the gradient descent method diverges for the specified learning rate of $\eta=0.1$.

### Comparison with Partial Differentiation

Using partial differentiation, set the derivative equal to zero:

$$
4(w_1-5)^3+2(w_1-3)=0
$$

Solving this equation gives the real critical point:

$$
\boxed{w_1\approx4.1649}
$$

This value is the minimum of the cost function.

The gradient descent method with the given learning rate:

$$
\eta=0.1
$$

did not converge to this minimum. Instead, the updates grew rapidly in magnitude and diverged.

Therefore, partial differentiation finds the minimum directly at approximately:

$$
\boxed{w_1=4.1649}
$$

while gradient descent with $\eta=0.1$ diverges because the learning rate is too large for this starting point and cost function.

## Question 5 – Linear Regression and Regularization

We want to minimize:

$$
\|y-Xw\|^2+\lambda\|w\|^2
$$

Using the L2 norm identity:

$$
\|v\|^2=v^Tv
$$

the objective function can be written as:

$$
(y-Xw)^T(y-Xw)+\lambda w^Tw
$$

Expanding the expression:

$$
J(w)
=
(y-Xw)^T(y-Xw)+\lambda w^Tw
$$

gives:

$$
J(w)
=
y^Ty-2w^TX^Ty+w^TX^TXw+\lambda w^Tw
$$

Take the gradient with respect to $w$:

$$
\nabla_w J
=
-2X^Ty+2X^TXw+2\lambda w
$$

Set the gradient equal to zero:

$$
-2X^Ty+2X^TXw+2\lambda w=0
$$

Divide by 2:

$$
-X^Ty+X^TXw+\lambda w=0
$$

Rearrange:

$$
X^TXw+\lambda w=X^Ty
$$

Factor out $w$:

$$
(X^TX+\lambda I)w=X^Ty
$$

Therefore:

$$
\boxed{
\hat w=(X^TX+\lambda I)^{-1}X^Ty
}
$$

Without regularization, the ordinary least-squares solution is:

$$
\hat w=(X^TX)^{-1}X^Ty
$$

With L2 regularization, the solution becomes:

$$
\hat w=(X^TX+\lambda I)^{-1}X^Ty
$$

The L2 penalty discourages the model from using very large weight values. As $\lambda$ increases, the weights are pushed closer to zero.

This can reduce overfitting and make the model more stable, although using too large a value of $\lambda$ can cause underfitting.

Therefore, the L2 penalty trades some model flexibility for improved stability and generalization.

## Question 6 – Assumptions in Linear Regression

The main assumptions of a simple linear regression model are:

1. **Linearity**  
   The relationship between the independent variable $x$ and the response variable $y$ is linear.

2. **Independence**  
   The observations and error terms are independent of one another.

3. **Zero mean of the errors**  
   The expected value of the error term is zero:

   $$
   E[\epsilon]=0
   $$

4. **Constant variance (Homoscedasticity)**  
   The variance of the error term is constant for all values of $x$:

   $$
   Var(\epsilon)=\sigma^2
   $$

5. **Normality of errors**  
   The error terms are normally distributed:

   $$
   \epsilon \sim N(0,\sigma^2)
   $$

6. **The independent variable is measured without error**  
   The values of $x$ are treated as known or fixed values rather than random measurement errors.

## Question 7 – Computing Derivative Using PyTorch Autograd

In [ ]:
import torch

r = torch.tensor(30000.1, dtype=torch.float64, requires_grad=True)
rs = torch.tensor(0.00887, dtype=torch.float64)

tau = 2 * r * torch.sqrt(r / rs) * torch.sqrt(1 - rs / r)

tau.backward()

print("tau =", tau.item())
print("dτ/dr =", r.grad.item())

tau = 110344877.04827532
dτ/dr = 5517.226005413796


Using PyTorch Autograd, the derivative at

$$
r=30000.1,\qquad r_s=0.00887
$$

is approximately:

$$
\boxed{\frac{d\tau}{dr}=5517.2260}
$$


## Question 8 – Linear Regression in One Variable with Scikit-Learn

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

url = "https://raw.githubusercontent.com/rahulbhadani/Datasets/master/RAV4/2020-04-15-05-12-00_2T3Y1RFV8KC014025_CAN_Messages_5_Hz.csv"

df = pd.read_csv(url)

X = df[["steer_angle"]]
y = df["yaw_rate"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)

model = LinearRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred)

print("Intercept:", model.intercept_)
print("Coefficient:", model.coef_[0])
print("R^2:", r2)

Intercept: -0.20719705434128108
Coefficient: 0.06620846316922513
R^2: 0.6555643761968142


### Explanation of Major Parts of the Code

The dataset is loaded into a pandas DataFrame using:

```python
df = pd.read_csv(url)
```

The independent variable is selected as the steering angle:

```python
X = df[["steer_angle"]]
```

and the response variable is selected as the yaw rate:

```python
y = df["yaw_rate"]
```

The data is then divided into training and testing sets using:

```python
train_test_split(X, y, test_size=0.25, random_state=42)
```

The value `test_size=0.25` means that 25% of the dataset is used for testing and the remaining 75% is used for training.

A linear regression model is created using:

```python
model = LinearRegression()
```

The model is trained using the training data:

```python
model.fit(X_train, y_train)
```

This finds the best-fit linear equation of the form:

$$
\hat{y}=w_0+w_1x
$$

where $w_0$ is the intercept and $w_1$ is the coefficient for steer angle.

Predictions are then made on the test data using:

```python
y_pred = model.predict(X_test)
```

The intercept and coefficient are obtained from:

```python
model.intercept_
model.coef_
```

Finally, the $R^2$ value is calculated using:

```python
r2_score(y_test, y_pred)
```

The $R^2$ value measures how much of the variation in yaw rate is explained by the linear relationship with steer angle on the test dataset.

## Question 9 – Linear Regression in One Variable with PyTorch

The goal is to create a custom linear regression model of the form:

$$
y=w_0+w_1x
$$

where:

- $x$ is the Benefit-Cost-Ratio (BCR)
- $y$ is AnnualProduction

The model will be implemented using PyTorch and gradient descent as part of the Python package.